music21/musicxml/m21ToXml.py

```diff
@@ -1672,6 +1743,186 @@ def setScoreHeader(self):
+            # Joinable groups must consist of only PartStaffs with Measures
+            if (
+                len(sg) > 1
+                and all(isinstance(p, stream.PartStaff) for p in sg)
+                and all(p.getElementsByClass('Measure') for p in sg)
+            ):
+                joinableGroups.append(sg)
+
+        # Pass 1: create <staff> tags and move elements
+        for group in joinableGroups:
+            initialRoot = None
+            for i, ps in enumerate(group):
+                staffNumber = i + 1  # 1-indexed
+                root = self._getRootForPartStaff(ps)
+
+                # Create <staff> tags under <note>, <direction>, <forward> tags
+                for mxMeasure in root.findall('measure'):
+                    XMLExporterBase.addStaffTag(mxMeasure, staffNumber,
+                        tagList=['note', 'direction', 'forward'])
+
+                # Move elements
+                if initialRoot is None:
+                    initialRoot = root
+                else:
+                    initialHigh = max(int(m.get('number')) for m in initialRoot.findall('measure'))
+                    thisHigh = max(int(m.get('number')) for m in root.findall('measure'))
+                    highestMeasureNumber = max(initialHigh, thisHigh)
+                    initialRootCursor = 0
+                    for mNum in range(highestMeasureNumber + 1):
+                        initialMeasure = initialRoot.find(f"measure[@number='{mNum}']")
+                        thisMeasure = root.find(f"measure[@number='{mNum}']")
+                        if thisMeasure is None and initialMeasure is None:
+                            # Gap in both measure sequences
+                            continue
+                        if thisMeasure is None:
+                            # Gap in this measure sequence
+                            initialRootCursor += 2  # Advance for comment & measure in initial seq.
+                            continue
+                        if initialMeasure is None:
+                            # Gap in initial part measure sequence, so insert entire measure
+                            divider = ET.Comment(DIVIDER_COMMENT.replace('n', str(mNum)))
+                            initialRoot.insert(initialRootCursor, divider)
+                            initialRootCursor += 1
+                            initialRoot.insert(initialRootCursor, thisMeasure)
+                            initialRootCursor += 1
+                            continue
+                        # No gaps found ...
+                        ScoreExporter.moveElements(thisMeasure, initialMeasure, staffNumber)
+                        initialRootCursor += 2  # comment & measure
+
+        # Pass 2: set number on initial clefs, measure attributes
+        # Need the earliest mxAttributes, which may not exist in initialRoot
+        # until moved there in Pass 1, e.g. RH of piano doesn't appear until m. 40
+        for group in joinableGroups:
```

Comment:
Do any of the joinableGroups affect the parts in other joinableGroups?  Could these be within the same loop as before but still with the two-part workflow?

Reply:
I suppose what I need to think through is whether there could be a Frankenstein set of `layout.StaffGroup` objects that would enclose the same `PartStaff` within multiple `StaffGroups` (and where those `StaffGroup`s, unlike orchestral scores, consist of all and only `PartStaffs`.) In essence, overlapping piano braces. I'm willing to think that through.

Reply:
Aaaand I got all of four seconds before realizing I've just articulated a scenario I can short-circuit and raise an exception for.

Reply:
On further thought, those kinds of overlapping StaffGroups might not be Frankenstein-y if handled carefully in some algorithmic composition use cases, so rather than rule them out, I can just check for errors closer to a problem source by raising exceptions in `addStaffTags()`, see https://github.com/cuthbertLab/music21/pull/634/commits/ce7039c0572cf1222c0e154278b581bf442f557c.